File: android/Reguerta/app/src/main/java/com/reguerta/user/data/profiles/FirestoreSharedProfileRepository.kt
package com.reguerta.user.data.profiles

import com.google.android.gms.tasks.Tasks
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.SetOptions
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.profiles.SharedProfile
import com.reguerta.user.domain.profiles.SharedProfileRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FirestoreSharedProfileRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : SharedProfileRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val profilesCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.SHARED_PROFILES)

    override suspend fun getAllSharedProfiles(): List<SharedProfile> = withContext(Dispatchers.IO) {
        try {
            val snapshot = Tasks.await(firestore.collection(profilesCollectionPath).get())
            snapshot.documents
                .map { document ->
                    val data: Map<String, Any> = document.data ?: invalidSharedProfileDocument()
                    decodeSharedProfileDocument(document.id, data)
                }
                .sortedByDescending(SharedProfile::updatedAtMillis)
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "sharedProfiles")
        }
    }

    override suspend fun getSharedProfile(userId: String): SharedProfile? = withContext(Dispatchers.IO) {
        try {
            val snapshot = Tasks.await(
                firestore.collection(profilesCollectionPath)
                    .document(userId)
                    .get(),
            )
            if (!snapshot.exists()) {
                null
            } else {
                val data: Map<String, Any> = snapshot.data ?: invalidSharedProfileDocument()
                decodeSharedProfileDocument(snapshot.id, data)
            }
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "sharedProfiles.document")
        }
    }

    override suspend fun upsertSharedProfile(profile: SharedProfile): SharedProfile = withContext(Dispatchers.IO) {
        try {
            Tasks.await(
                firestore.collection(profilesCollectionPath)
                    .document(profile.userId)
                    .set(sharedProfileUpsertPayload(profile), SetOptions.merge()),
            )
            profile
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "sharedProfiles.write")
        }
    }

    override suspend fun deleteSharedProfile(userId: String): Boolean = withContext(Dispatchers.IO) {
        try {
            Tasks.await(
                firestore.collection(profilesCollectionPath)
                    .document(userId)
                    .delete(),
            )
            true
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "sharedProfiles.write")
        }
    }
}

internal fun sharedProfileUpsertPayload(profile: SharedProfile): Map<String, Any> = mapOf(
    "userId" to profile.userId,
    "familyNames" to profile.familyNames,
    "photoUrl" to (profile.photoUrl ?: FieldValue.delete()),
    "about" to profile.about,
    "updatedAt" to Timestamp(
        profile.updatedAtMillis / 1_000,
        ((profile.updatedAtMillis % 1_000) * 1_000_000).toInt(),
    ),
)

internal fun decodeSharedProfileDocument(
    documentId: String,
    data: Map<String, Any?>,
): SharedProfile {
    val normalizedDocumentId = documentId.trim()
    if (normalizedDocumentId.isBlank() ||
        data.requiredString("userId") != normalizedDocumentId
    ) {
        invalidSharedProfileDocument()
    }
    val updatedAt = data["updatedAt"] as? Timestamp ?: invalidSharedProfileDocument()
    return SharedProfile(
        userId = normalizedDocumentId,
        familyNames = data.optionalString("familyNames").orEmpty(),
        photoUrl = data.optionalString("photoUrl"),
        about = data.optionalString("about").orEmpty(),
        updatedAtMillis = updatedAt.toDate().time,
    )
}

private fun Map<String, Any?>.requiredString(field: String): String =
    optionalString(field) ?: invalidSharedProfileDocument()

private fun Map<String, Any?>.optionalString(field: String): String? {
    val rawValue = this[field] ?: return null
    val value = rawValue as? String ?: invalidSharedProfileDocument()
    return value.trim().ifBlank { null }
}

private fun invalidSharedProfileDocument(): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = "sharedProfiles.document",
)

File: ios/Reguerta/Reguerta/Data/Profiles/FirestoreSharedProfileRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

actor FirestoreSharedProfileRepository: SharedProfileRepository {
    private let storedDB: Firestore

    init(firebaseAppName: String) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for shared profiles")
        }
        self.storedDB = Firestore.firestore(app: app)
    }

    func allSharedProfiles(environment: SessionEnvironment) async throws -> [SharedProfile] {
        do {
            let profilesCollection = storedDB.reguertaCollection(.sharedProfiles, environment: environment)
            try Task.checkCancellation()
            let snapshot = try await profilesCollection.getDocuments()
            return try snapshot.documents
                .map { document in
                    try Self.sharedProfile(
                        documentID: document.documentID,
                        data: document.data()
                    )
                }
                .sorted { $0.updatedAtMillis > $1.updatedAtMillis }
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "sharedProfiles")
        }
    }

    func sharedProfile(userId: String, environment: SessionEnvironment) async throws -> SharedProfile? {
        do {
            let profilesCollection = storedDB.reguertaCollection(.sharedProfiles, environment: environment)
            try Task.checkCancellation()
            let document = try await profilesCollection.document(userId).getDocument()
            guard document.exists else { return nil }
            guard let data = document.data() else { throw Self.invalidDocumentError }
            return try Self.sharedProfile(documentID: document.documentID, data: data)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "sharedProfiles.document")
        }
    }

    func upsert(profile: SharedProfile, environment: SessionEnvironment) async throws -> SharedProfile {
        do {
            let profilesCollection = storedDB.reguertaCollection(.sharedProfiles, environment: environment)
            try Task.checkCancellation()
            try await profilesCollection.document(profile.userId).setData(
                Self.upsertPayload(for: profile),
                merge: true
            )
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "sharedProfiles.write")
        }
        return profile
    }

    func deleteSharedProfile(userId: String, environment: SessionEnvironment) async throws -> Bool {
        do {
            let profilesCollection = storedDB.reguertaCollection(.sharedProfiles, environment: environment)
            try Task.checkCancellation()
            try await profilesCollection.document(userId).delete()
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "sharedProfiles.write")
        }
        return true
    }

    static func upsertPayload(for profile: SharedProfile) -> [String: Any] {
        [
            "userId": profile.userId,
            "familyNames": profile.familyNames,
            "photoUrl": firestoreValue(profile.photoUrl),
            "about": profile.about,
            "updatedAt": Timestamp(date: Date(timeIntervalSince1970: TimeInterval(profile.updatedAtMillis) / 1_000))
        ]
    }

    static func sharedProfile(documentID: String, data: [String: Any]) throws -> SharedProfile {
        let normalizedDocumentID = documentID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedDocumentID.isEmpty,
              try requiredString(data, field: "userId") == normalizedDocumentID,
              let updatedAt = data["updatedAt"] as? Timestamp else {
            throw invalidDocumentError
        }

        return SharedProfile(
            userId: normalizedDocumentID,
            familyNames: try optionalString(data, field: "familyNames") ?? "",
            photoUrl: try optionalString(data, field: "photoUrl"),
            about: try optionalString(data, field: "about") ?? "",
            updatedAtMillis: Int64(updatedAt.dateValue().timeIntervalSince1970 * 1_000)
        )
    }

    private static func requiredString(_ data: [String: Any], field: String) throws -> String {
        guard let value = try optionalString(data, field: field) else { throw invalidDocumentError }
        return value
    }

    private static func optionalString(_ data: [String: Any], field: String) throws -> String? {
        guard let rawValue = data[field], !(rawValue is NSNull) else { return nil }
        guard let value = rawValue as? String else { throw invalidDocumentError }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return normalized.isEmpty ? nil : normalized
    }

    private static func firestoreValue<Value>(_ value: Value?) -> Any {
        if let value {
            return value
        }
        return FieldValue.delete()
    }

    private static var invalidDocumentError: RepositoryError {
        .invalidData(resource: "sharedProfiles.document")
    }
}
